### ISSUE TITLE
State machines

### ISSUE THREAD

**Issue by gfrison** [jenetics/jenetics#352]
Hello all!

I would like to ask you whether it is possible to state machines with jenetics.
for example, I would provide a fitness function which check the result as the sum of all elements added so far, plus the integer provided in that step.
step 1:
1 -> 1
2 -> 3
5 -> 8
....

how can I implement this with jenetics?

Thank you

**Comment by jenetics** [comment]
I'm not sure what you are trying to achieve. Do you want to evolve a state-machine? Do you want to evolve the state-machine-graph or _just_ the transition-event parameters. I also didn't get how you are going to calculate the fitness of your state-machine.
A little bit context would be helpful.

**Comment by gfrison** [comment]
I'm wondering how could be possible to implement something like Mealy machine with jenetics.
In particular, I'm interested on a a system which for a given input returns the output and the new state of the program that can handle the following input.
let assume InputN are the inputs on step n, likewise OutputN (Program) are the outputs (GA states) on step n.
Program0(Input0) -> (Output0,  Program1)

In the case of the previous example, the GA sums the input with the sum of previous inputs:
step N: InputN -> OutputN
step 1: 1 -> 1
step 2: 3 -> 4     (1+3)
step 3: 2 -> 6     (4+2)
basically, the GA keep the memory of the current state (changing its choromosome).
I see you are working on a timeseries branch, maybe it is exactly what I am looking for.
Is it possible to implement that finite state machine with jenetics?

Thank you so much!

**Comment by jenetics** [comment]
I think there is  a slight misunderstanding on how Jenetics works. The GA itself doesn't store any state. What you need is a codec which maps the `Genotype` to an instance of your `StateMachine`. The `StateMachine` represents the topology _and_ the initial internal state. Your fitness function will than execute the state-machine and calculate it's fitness. The GA then tries to evolve state-machine with better fitness.
So, I would start to implement the `StateMachine` class and a proper mapping to a `Genotype`. Once you have the `StateMachine` in place I can try to help you to find a `Genotype` mapping.

**Comment by gfrison** [comment]
hello,

what I mean it's something like [recurrent cartesian GP](https://www.researchgate.net/publication/277929205_Recurrent_Cartesian_Genetic_Programming), how could be possible to have it in jenetics? 
Or at least, what are the parts that should be added for implementing RCGP?

thank you

**Comment by gfrison** [comment]
Hello,

I implemented a stateful ``Op`` which return the value with a 'delay'. it is here https://gist.github.com/gfrison/276b4a50c1c0acfa8019c9b5d2b0d5ab with the class ``CacheOp`` it should replicate the results of the recurrent cartesian GP previously indicated. in the same gist, I implemented a simple ``ProgramGene`` which return the sum of all past values plus the current given one. 
With very simple configuration, indicating only 2 available operations (ADD, Cache) sometimes it turns out correctly:
```
gene:add(x, cache(add(cache(add(cache(add(cache(x), x)), x)), x)))
n:88, predict:88, actual:88
n:28, predict:116, actual:116
n:92, predict:208, actual:208
n:1, predict:209, actual:209
n:7, predict:216, actual:216
n:57, predict:185, actual:273
n:64, predict:221, actual:337
```
by displacing correctly ADD and cache one after another. the problem is when it comes to handle a bigger solutions spaces, such as with more math operations or longer sequences to manage.
Do you think the snippet I created is correctly configured for the type of problem?

Thank you

**Comment by jenetics** [comment]
Implementing a _stateful_ `Op` is against the _spirit_ of the `jenetics.prog` module. I don't think it can work reliable, since it depends on the caching strategy of the fitness value of the  `Phenotype` class. The caching in the `Phenotype` has been implemented under the assumption, that the gene values will not change. This is not the case for your `CacheOp`.
